Firefox browser install now uses mozdownload and mozinstall

Cactusmachete opened the pull request:
mozinstall will refuse to install if firefox is already installed
(when checked on OSX). I've currently only set it to print an error message
when this happens - it'll require the user to uninstall firefox in _venv and run
the command again as uninstall in mozinstall doesn't seem to work very well.

Closes https://github.com/w3c/web-platform-tests/issues/7996

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/9748)
<!-- Reviewable:end -->

Cactusmachete committed:
Firefox browser install now uses mozdownload and mozinstall

mozinstall will refuse to install if firefox is already installed
(when checked on OSX). I've currently only set it to print an error message
when this happens - it'll require the user to uninstall firefox in _venv and run
the command again as uninstall in mozinstall doesn't seem to work very well.

Closes https://github.com/w3c/web-platform-tests/issues/7996

jgraham commented on tools/wpt/browser.py:
```diff
@@ -105,36 +107,23 @@ def platform_string_geckodriver(self):
-        resp = get(nightly_link)
-        resp.raise_for_status()
-        untar(resp.raw, dest=dest)
+        filename = FactoryScraper('daily', branch='mozilla-central').download()
```
I think we should set a destination here so that it ends up inside the venv directory.

jgraham commented on tools/wpt/browser.py:
```diff
@@ -105,36 +107,23 @@ def platform_string_geckodriver(self):
+        try:
+            #When tested in OSX, mozinstall will refuse to install if firefox is already installed
+            mozinstall.install(filename, dest)
+        except Exception as e:
```
It would be nice to catch a more specific exception type here.

jgraham commented on tools/wpt/browser.py:
```diff
@@ -105,36 +107,23 @@ def platform_string_geckodriver(self):
+            #When tested in OSX, mozinstall will refuse to install if firefox is already installed
+            mozinstall.install(filename, dest)
+        except Exception as e:
+            # Instead of printing an error message we should uninstall the currently
```
Can we just figure out what to delete based on the path, or does something more complex happen on OSX here?

Cactusmachete replied:
Messed around a bit, turns out it's a simple fix. I still haven't tested this on windows though, which might require a different uninstall path. Will look into that in a few hours.

jgraham reviewed (changes requested):
This is a very exciting change; I love all the deleted code :) I think a little cleanup is required to land it though.

w3c-bots commented:
# Build BROKEN

Started: 2018-03-02 18:14:44
Finished: 2018-03-02 18:21:50


<h2>Failing Jobs</h2>
<ul>

<li>wpt_integration in py27,py27-flake8</li>

</ul>




View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/25151)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/348337460)

Cactusmachete committed:
Added support for uninstalling on OSX

Wasn't sure why this was failing on travis. Added mozdownload and mozinstall
to firefox_requirements hoping that would help.

Cactusmachete committed:
Removed test_browser.py as it was causing travis ci to fail

test_browser was calling dead code in browser.py, so travis failed.

jgraham commented on tools/wptrunner/requirements_firefox.txt:
```diff
@@ -4,3 +4,5 @@ mozprocess == 0.26
 mozcrash == 1.0
 mozrunner == 6.14
 mozleak == 0.1
+mozinstall == 1.15
```
This should't be both in here and in the commands.json file; here is better but you might need to move the `moz*` imports you added into ttthe function where they're actually used.

Cactusmachete replied:
If I remove these requirements from the commands.json file, the code crashes when I try to set up a new venv because mozdownload/install don't get installed there in it.
When I remove them from this file, it could cause the build to fail in Travis if anything calls browse.py because they don't get installed there.

jgraham replied:
So I think this is possible, but it involves ensuring that we install the product-specific requriements later on. Let's do that in a followup.

jgraham commented on tools/wpt/browser.py:
```diff
@@ -105,36 +107,30 @@ def platform_string_geckodriver(self):
+                "Darwin": "mac"
+            }.get(uname[0])
+
+            if platform == "mac":
```
`if uname[0] == "Darwin"` seems simpler if we really only need this on one platform. But it's unclear why macOS is special cased here. Does it really work differently on mac? If it does a comment explaining what's going on seems helpful, otherwise we should support all for Windows/Linux/Mac here

Cactusmachete replied:
Alright!
I checked it on all three OSes. mozinstall only throws an error on Mac. 
Afaict this is because mozdownload gets a  .dmg on Mac, but zipped folders in Lin/Win which it simply unzips there.

jgraham commented on tools/wpt/browser.py:
```diff
@@ -105,36 +107,30 @@ def platform_string_geckodriver(self):
+                raise
+
+        os.remove(filename)
         return find_executable("firefox", os.path.join(dest, "firefox"))
```
Should we consider `mozinstall.get_binary` here?

Cactusmachete replied:
We could, but I have no way of testing this on a Mac anymore.

Cactusmachete committed:
Added comments

Cactusmachete committed:
removed time import

csnardi commented:
It looks like this broke CI, e.g. https://travis-ci.org/w3c/web-platform-tests/jobs/349449358.
